pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java

```diff
@@ -217,7 +223,18 @@ public void reload(Properties properties) throws Exception {
     }
 
     public void run() throws Exception {
-        final Terminal terminal = TerminalBuilder.builder().build();
+        final Terminal terminal = TerminalBuilder.builder()
+                .nativeSignals(true)
+                .signalHandler(signal -> {
+                    if (signal == Terminal.Signal.INT || signal == Terminal.Signal.QUIT) {
+                        if (execState == ExecState.RUNNING) {
+                            throw new InterruptShellException();
```

Comment:
In which thread is this exception thrown to?

Since we swallow `InterruptShellException` below, I'm unsure how the shell exits after the RUNNING command finished.

Reply:
This class is mono-thread

If you double click control+c the shell exits because the first one makes the running command terminate and the second one exit the shell (as before)

the command termination one is handled here: https://github.com/apache/pulsar/blob/f8e632a45eece08d5c3454c6681484f02c0150cb/pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java#L453-L457

the shell termination is here:

https://github.com/apache/pulsar/blob/f8e632a45eece08d5c3454c6681484f02c0150cb/pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java#L424-L431